diff -crB vlc-0.9.9_ori/modules/codec/vorbis.c vlc-0.9.9_mod/modules/codec/vorbis.c
*** vlc-0.9.9_ori/modules/codec/vorbis.c	2008-08-28 06:32:51.000000000 +0800
--- vlc-0.9.9_mod/modules/codec/vorbis.c	2012-09-21 17:45:22.494713747 +0800
***************
*** 537,547 ****
  #endif
 
      if( p_oggpacket->bytes &&
- #ifdef MODULE_NAME_IS_tremor
-         vorbis_synthesis( &p_sys->vb, p_oggpacket, 1 ) == 0 )
- #else
          vorbis_synthesis( &p_sys->vb, p_oggpacket ) == 0 )
- #endif
          vorbis_synthesis_blockin( &p_sys->vd, &p_sys->vb );

      /* **pp_pcm is a multichannel float vector. In stereo, for
--- 537,543 ----
