diff -Nur a/modules/codec/avcodec/audio.c b/modules/codec/avcodec/audio.c
--- a/modules/codec/avcodec/audio.c	2008-09-11 08:04:17.000000000 +0800
+++ b/modules/codec/avcodec/audio.c	2013-01-10 14:44:00.234719593 +0800
@@ -250,6 +250,17 @@
 
     return p_buffer;
 }
+int  avcodec_decode_audio2(AVCodecContext *avctx, int16_t *samples,
+                         int *frame_size_ptr,
+                         const uint8_t *buf, int buf_size)
+{
+    AVPacket avpkt;
+    av_init_packet(&avpkt);
+    avpkt.data = buf;
+    avpkt.size = buf_size;
+
+    return avcodec_decode_audio3(avctx, samples, frame_size_ptr, &avpkt);
+}
 
 /*****************************************************************************
  * DecodeAudio: Called to decode one frame
