diff -Nur a/m4/rygel.m4 rygel-0.14.1/m4/rygel.m4
--- a/m4/rygel.m4	2013-01-24 15:41:45.087402802 +0800
+++ rygel-0.14.1/m4/rygel.m4	2013-01-24 15:42:59.707405435 +0800
@@ -99,6 +99,15 @@
           [dnl check for vala
            AM_PROG_VALAC([$1])
 
+           for var in vapidir vapidir_versioned; do
+                AS_IF([dir=`$PKG_CONFIG --variable=$var vapigen`],
+                    [
+                        AC_MSG_NOTICE([got $var=$dir in ${datadir}from pkg-config])
+                        RYGEL_ADD_VALAFLAGS([--vapidir ${STAGING_PREFIX}/../$dir])
+                    ],
+                    [AC_MSG_NOTICE([pkg-config didn't return a value for $var, relying on valac built-in defaults])])
+            done
+
             AS_IF([test x$VALAC = "x"],
                 [AC_MSG_ERROR([Cannot find the "valac" compiler in your PATH])],
                 [
