*** rygel-0.4.2/src/plugins/gst-renderer/rygel-gst-connection-manager.vala	2012-11-14 08:53:25.278888602 +0800
--- rygel-0.4.2_mod/src/plugins/gst-renderer/rygel-gst-connection-manager.vala	2012-11-14 08:52:48.693595643 +0800
***************
*** 46,56 ****
--- 46,61 ----
                                      "http-get:*:audio/x-wav:*," +
  									"http-get:*:audio/x-vorbis:*," +
                                      "http-get:*:audio/x-vorbis+ogg:*," +
+ 									"http-get:*:audio/ogg:*," +
+ 									"http-get:*:application/ogg:*," +
  									"http-get:*:audio/x-ms-wma:*," +
                                      "http-get:*:audio/x-ms-asf:*," +
                                      "http-get:*:audio/x-flac:*," +
+ 									"http-get:*:audio/x-flac+ogg:*," +
+ 									"http-get:*:audio/flac:*," +
                                      "http-get:*:audio/x-mod:*," +
  									"http-get:*:audio/x-ac3:*," +
+ 									"http-get:*:audio/mp4a-latm:*," +
                                      "http-get:*:audio/x-m4a:*";
      }
  
