--- a/src/plugins/gst-renderer/rygel-gst-connection-manager.vala
+++ b/src/plugins/gst-renderer/rygel-gst-connection-manager.vala
@@ -46,11 +46,16 @@ public class Rygel.GstConnectionManager
                                     "http-get:*:audio/x-wav:*," +
 									"http-get:*:audio/x-vorbis:*," +
                                     "http-get:*:audio/x-vorbis+ogg:*," +
+ 									"http-get:*:audio/ogg:*," +
+ 									"http-get:*:application/ogg:*," +
 									"http-get:*:audio/x-ms-wma:*," +
                                     "http-get:*:audio/x-ms-asf:*," +
                                     "http-get:*:audio/x-flac:*," +
+ 									"http-get:*:audio/x-flac+ogg:*," +
+ 									"http-get:*:audio/flac:*," +
                                     "http-get:*:audio/x-mod:*," +
 									"http-get:*:audio/x-ac3:*," +
+ 									"http-get:*:audio/mp4a-latm:*," +
                                     "http-get:*:audio/x-m4a:*";
     }
 
