diff -Nur a/linux/net/ipsec/ipsec_rcv.c openswan-2.6.38/linux/net/ipsec/ipsec_rcv.c
--- a/linux/net/ipsec/ipsec_rcv.c	2014-07-01 15:32:26.766900956 +0800
+++ openswan-2.6.38/linux/net/ipsec/ipsec_rcv.c	2014-07-01 16:41:40.626978939 +0800
@@ -955,7 +955,10 @@
 
 		for(i = 0; i <= ipsecdevices_max; i++) {
 			if(ipsecdevices[i] == NULL) continue;
-			if(!netdev_priv(ipsecdevices[i])) continue;
+            /*When ipsec_tunnel is not finished, the magic won't be filled with right value
+             *it will cause a bug report. check the type before check the magci
+             */
+			if(ipsecdevices[i]->type != 31 || !netdev_priv(ipsecdevices[i])) continue;
 
 			prvdev = netdev_to_ipsecpriv(ipsecdevices[i]);
 			
